File: data-structures/queue/problems/ts/leetcode-933.ts
class RecentCounter {
    private queue: Array<number>
    private head: number
    private tail: number
    constructor() {
        this.queue = new Array(0)
        this.head = 0
        this.tail = 0
    }

    ping(t: number): number {
        this.queue.push(t)
        this.tail = this.queue.length - 1
        return this.calculateSum()
    }

    private calculateSum(): number {
        const last = this.queue[this.tail]
        const searchSpace = last - 3000

        // console.log(this.queue[this.head])
        while(this.queue[this.head] < searchSpace) {
            this.queue.shift()
            // console.log(this.queue)
        }

        return this.queue.length
    }
}

const rc = new RecentCounter()
console.log(rc.ping(1))
console.log(rc.ping(100))
console.log(rc.ping(3001))
console.log(rc.ping(3002))

File: data-structures/queue/problems/java/leetcode-933.java
import java.util.ArrayDeque;
import java.util.Queue;

class RecentCounter {
    private Queue<Integer> queue;
    private int tail;

    public RecentCounter() {
        this.queue = new ArrayDeque<Integer>();
        this.tail = 0;
    }

    public int ping(int t) {
        this.queue.add(t);
        this.tail = t;

        return this.calculateSum();
    }

    private int calculateSum() {
        if (this.queue.peek() == null) {
            return 0;
        }
        Integer last = this.tail;

        int searchSpace = last - 3000;
        while (searchSpace > this.queue.peek()) {
            this.queue.poll();
        }

        return this.queue.size();
    }
}